middle_end/flambda2/types/meet_and_join.ml

```diff
@@ -46,22 +32,321 @@ let all_aliases_of env simple_opt ~in_env =
+      match res.all_left, res.all_right with
+      | true, true -> Both_inputs
+      | true, false -> Left_input
+      | false, true -> Right_input
+      | false, false -> New_result (rebuild res.mapping)
+    in
+    Ok (result, res.env)
+  with Bottom_result -> Bottom
+
+module Map_meet (M : Container_types_intf.Map) = struct
+  let meet ~(meet_data : TE.t -> 'a -> 'a -> 'a meet_result) env (left : 'a M.t)
+      (right : 'a M.t) : 'a M.t meet_result =
+    let fold2 f m1 m2 init =
+      let r = ref init in
+      let _m =
+        M.merge
+          (fun key left right ->
+            r := f key left right !r;
+            None)
+          m1 m2
+      in
+      !r
+    in
+    let rebuild l =
+      List.fold_left (fun m (key, data) -> M.add key data m) M.empty l
+    in
+    let fold2 = { fold2 } in
+    meet_mapping ~meet_data ~fold2 ~env ~left ~right ~rebuild
+end
+
+module Function_slot_map_meet = Map_meet (Function_slot.Map)
+module Value_slot_map_meet = Map_meet (Value_slot.Map)
+
+type 'a pairwise_disjunction_meet_arg =
+  { is_bottom : 'a -> bool;
+    mk_bottom : unit -> 'a;
+    meet : TE.t -> 'a -> 'a -> 'a meet_result
+  }
+
+let pairwise_disjunction_meet ~join_env_extension ~meet_type meetx meety
+    initial_env x1 x2 y1 y2 =
+  let join_scope = TE.current_scope initial_env in
+  let env = TE.increment_scope initial_env in
+  let to_extension scoped_env =
+    TE.cut scoped_env ~cut_after:join_scope
+    |> Typing_env_level.as_extension_without_bindings
+  in
+  let direct_return r =
+    map_env r ~f:(fun scoped_env ->
+        TE.add_env_extension_strict initial_env (to_extension scoped_env)
+          ~meet_type)
+  in
+  match meetx.meet env x1 x2, meety.meet env y1 y2 with
+  | Bottom, Bottom -> Bottom
+  | Ok (xr, env), Bottom ->
+    let x = extract_value xr x1 x2 in
+    if meetx.is_bottom x
+    then Bottom
+    else
+      let y = meety.mk_bottom () in
+      direct_return (Ok (New_result (x, y), env))
+  | Bottom, Ok (yr, env) ->
+    let y = extract_value yr y1 y2 in
+    if meety.is_bottom y
+    then Bottom
+    else
+      let x = meetx.mk_bottom () in
+      direct_return (Ok (New_result (x, y), env))
+  | Ok (xr, xenv), Ok (yr, yenv) ->
+    let x = extract_value xr x1 x2 in
+    let y = extract_value yr y1 y2 in
+    let bottom_x = meetx.is_bottom x in
+    let bottom_y = meety.is_bottom y in
+    let result_env _ : _ Or_bottom.t =
+      let[@local] add_extension ext =
+        TE.add_env_extension_strict initial_env ext ~meet_type
+      in
+      if bottom_x
+      then if bottom_y then Bottom else add_extension (to_extension yenv)
+      else if bottom_y
+      then add_extension (to_extension xenv)
+      else
+        let x_extension = to_extension xenv in
+        let y_extension = to_extension yenv in
+        let join_env = Join_env.create env ~left_env:xenv ~right_env:yenv in
+        let extension = join_env_extension join_env x_extension y_extension in
+        add_extension extension
```

Comment:
Those cases doing `to_extension` `add_extension` might be quite expensive:
Adding those extensions to the environment might have to basically completely redo the meet work (by calling back meet_type)
When doing a meet between two deep (depth n) variant types producing m equations, this might do some work of size `n * m`

Reply:
Given that the environment to which we want to add the extensions is the direct parent of the one from which we extract the extension, there are probably more efficient way to do that.

If the assumptions that extensions from the top levels are more precise than the deeper ones, we could directly replace them (maybe).

We probably don't want to do that assumption here: we might want to add some function to collapse the last level from the typing env